packages/controllers/src/subject-metadata/SubjectMetadataController.test.ts

```diff
@@ -0,0 +1,228 @@
+    'PermissionController:hasPermissions',
+    hasPermissionsSpy,
+  );
+
+  return [
+    controllerMessenger.getRestricted<
+      typeof controllerName,
+      SubjectMetadataControllerActions['type'] | HasPermissions['type'],
+      SubjectMetadataControllerEvents['type']
+    >({
+      name: controllerName,
+      allowedActions: [
+        'PermissionController:hasPermissions',
+        'SubjectMetadataController:getState',
+      ],
+    }),
+    hasPermissionsSpy,
+  ] as const;
+}
+
+function getSubjectMetadata(
+  origin: string,
+  name: string,
+  opts?: Record<string, Json>,
+) {
+  return {
+    origin,
+    name,
+    iconUrl: null,
+    extensionId: null,
+    ...opts,
+  };
+}
+
+describe('SubjectMetadataController', () => {
+  describe('constructor', () => {
+    it('initializes a subject metadata controller', () => {
+      const controller = new SubjectMetadataController({
+        messenger: getSubjectMetadataControllerMessenger()[0],
+        subjectCacheLimit: 10,
+      });
+      expect(controller.state).toStrictEqual({ subjectMetadata: {} });
+    });
+
+    it('trims subject metadata state on startup', () => {
+      const [messenger, hasPermissionsSpy] =
+        getSubjectMetadataControllerMessenger();
+      hasPermissionsSpy.mockImplementationOnce(() => false);
+      hasPermissionsSpy.mockImplementationOnce(() => true);
+
+      const controller = new SubjectMetadataController({
+        messenger,
+        subjectCacheLimit: 10,
+        state: {
+          subjectMetadata: {
+            'foo.com': getSubjectMetadata('foo.com', 'foo'),
+            'bar.io': getSubjectMetadata('bar.io', 'bar'),
+          },
+        },
+      });
+
+      expect(controller.state).toStrictEqual({
+        subjectMetadata: { 'bar.io': getSubjectMetadata('bar.io', 'bar') },
+      });
+    });
+
+    it('throws if the subject cache limit is invalid', () => {
+      [0, -1, 1.1].forEach((subjectCacheLimit) => {
+        expect(
+          () =>
+            new SubjectMetadataController({
+              messenger: getSubjectMetadataControllerMessenger()[0],
+              subjectCacheLimit,
+            }),
+        ).toThrow(
+          `subjectCacheLimit must be a positive integer. Received: "${subjectCacheLimit}"`,
+        );
+      });
+    });
+  });
+
+  describe('clearState', () => {
+    it('clears the controller state', () => {
+      const [messenger, hasPermissionsSpy] =
+        getSubjectMetadataControllerMessenger();
+      hasPermissionsSpy.mockImplementation(() => true);
+
+      const controller = new SubjectMetadataController({
+        messenger,
+        subjectCacheLimit: 10,
+      });
+
+      controller.addSubjectMetadata(getSubjectMetadata('foo.com', 'foo'));
+
+      expect(controller.state).toStrictEqual({
+        subjectMetadata: { 'foo.com': getSubjectMetadata('foo.com', 'foo') },
+      });
+
+      expect(
+        (controller as any).subjectsEncounteredSinceStartup.size,
```

Comment:
This test is looking at internals instead of asserting what it's meant to prove, which is that the trimming still works correctly.

Instead of doing this, you could test that after clearing state, it's capable of holding X new sites again, where X is the limit. That would confirm this aspect of clearing state works correctly, and it would be less coupled with the implementation.

Reply:
Ahhh, I had to try.

Reply:
Fixed here: 66d6eb5